src/diffusers/models/controlnet_xs.py

```diff
@@ -0,0 +1,1060 @@
+        cross_attention_dim (`int`, defaults to 1024):
+            The dimension of the cross attention features.
+        down_block_types (`list[str]`, defaults to `["CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "DownBlock2D"]`):
+            The tuple of downsample blocks to use.
+        sample_size (`int`, defaults to 96):
+            Height and width of input/output sample.
+        transformer_layers_per_block (`Union[int, Tuple[int]]`, defaults to 1):
+            The number of transformer blocks of type [`~models.attention.BasicTransformerBlock`]. Only relevant for
+            [`~models.unet_2d_blocks.CrossAttnDownBlock2D`], [`~models.unet_2d_blocks.UNetMidBlock2DCrossAttn`].
+        upcast_attention (`bool`, defaults to `True`):
+            Whether the attention computation should always be upcasted.
+        max_norm_num_groups (`int`, defaults to 32):
+            Maximum number of groups in group normal. The actual number will the the largest divisor of the respective channels, that is <= max_norm_num_groups.
+    """
+
+    @staticmethod
+    def gather_base_subblock_sizes(blocks_sizes: List[int]):
+        """
+        To create a correctly sized `ControlNetXSAddon`, we need to know
+        the channels sizes of each base subblock.
+
+        Parameters:
+            blocks_sizes (`List[int]`):
+                Channel sizes of each base block.
+        """
+
+        n_blocks = len(blocks_sizes)
+        n_subblocks_per_block = 3
+
+        down_out = []
+        up_in = []
+
+        # down_out
+        for b in range(n_blocks):
+            for i in range(n_subblocks_per_block):
+                if b == n_blocks - 1 and i == 2:
+                    # Last block has no downsampler, so there are only 2 subblocks instead of 3
+                    continue
+
+                # The input channels are changed by the first resnet, which is in the first subblock.
+                if i == 0:
+                    # Same input channels
+                    down_out.append(blocks_sizes[max(b - 1, 0)])
+                else:
+                    # Changed input channels
+                    down_out.append(blocks_sizes[b])
+
+        down_out.append(blocks_sizes[-1])
+
+        # up_in
+        rev_blocks_sizes = list(reversed(blocks_sizes))
+        for b in range(len(rev_blocks_sizes)):
+            for i in range(n_subblocks_per_block):
+                # The input channels are changed by the first resnet, which is in the first subblock.
+                if i == 0:
+                    # Same input channels
+                    up_in.append(rev_blocks_sizes[max(b - 1, 0)])
+                else:
+                    # Changed input channels
+                    up_in.append(rev_blocks_sizes[b])
+
+        return {
+            "down - out": down_out,
+            "mid - out": blocks_sizes[-1],
+            "up - in": up_in,
+        }
+
+    @classmethod
+    def from_unet(
```

Comment:
do we need this?

Reply:
I think so, as the shape of a `ControlNetXSAddon` depends on the `UNet` it is for. Therefore users will use this method to create a `ControlNetXSModel` like this:

```python
from diffusers import UNet2DConditionModel, ControlNetXSAddon, ControlNetXSAddon

unet = UNet2DConditionModel.from_pretrained(...)
controlnet_xs_addon = ControlNetXSAddon.from_unet(unet, size_ratio=0.1)
controlnet_xs = ControlNetXSModel(unet, controlnet_xs_addon)

# train controlnet_xs
```

Reply:
so this is for training only, right?

Reply:
Yes